### DISCUSSION TITLE
Terraform re-apply is inconsistant while trying to update the existing environment which is provisioned from the same code.

### DISCUSSION THREAD

**Discussion by railoni** [jfrog/terraform-provider-artifactory#872] [Q&A]
Terraform re-apply is inconsistent while trying to update the existing environment which is provisioned from the same code.
I am using the below providers to configure artifactory with creating projects, users, repositories etc.

```terraform
terraform {
  required_providers {
    artifactory = {
      source  = "jfrog/artifactory"
      version = "9.7.0"
    }
    project = {
      source  = "jfrog/project"
      version = "1.3.3"
    }
    xray = {
      source  = "jfrog/xray"
      version = "2.0.1"
    }
    random = {
      source = "hashicorp/random"
      version = "3.5.1"
    }
  }
}
```
Every re-apply environment getting empty for all the repos. can someone help me to fix and have consistent for every run.

**Comment by alexhung** [comment]
Hi @railoni, can you provide more information on your issue? What version of JFrog products (Artifactory, Xray, etc.) are you using? What is the rest of your configuration? Which resources are you using? What are the output from running `terraform apply`?

**Reply by railoni** [reply]
Hi @alexhung Thank you for the promt response.
We are using jfrog artifactory Enterprise Plus 7.68.7 version and Xray Version 3.82.6.
Using resource "artifactory_federated_maven_repository" for creating fedrated repos It's not setting the project_environments            = ["PROD"]
so i have created local-exec to set the project environment.
 provisioner "local-exec" {
    command = <<-EOT
      exec curl -u 'admin:${var.ha2_password}' -X POST ${var.artifactory-ha2_url}artifactory/api/repositories/${project.BU_project_HA2.key}-maven-remote-prod -H 'Content-Type: application/json' -d '{"environments":["PROD"]}'
    EOT
  }
  For the first run it's working as expected but from the next apply project environment getting removed. i will re-run and share the output.

**Reply by railoni** [reply]
hi @alexhung could you please review the output file and let us know your inputs on fixing our issue.
[output.txt](https://github.com/jfrog/terraform-provider-artifactory/files/14028424/output.txt)

**Reply by railoni** [reply]
here is the re-apply refresh output.txt which just deleted the project for all the repos.
[output.txt](https://github.com/jfrog/terraform-provider-artifactory/files/14028457/output.txt)

**Reply by danielmkn** [reply]
Hi @railoni, Alex will be back next week and will take a look.

**Reply by railoni** [reply]
Thank you letting me know @danielmkn

**Reply by railoni** [reply]
Hi @alexhung if you are back from vacation. Could you please help me with the changes need to be done on tf code to fix inconsistency.

**Reply by alexhung** [reply]
@railoni From your output file, I don't see the configuration for your repos. I assume they exist. Do they have the `project_key` attribute set to the `key` attribute from the corresponding `project` resource?

You can use `lifecycle.ignore_changes` meta attribute on your repo configuration to ignore the state drift:

```terraform
lifecycle {
  ignore_changes = [
    project_key
  ]
}
```

Also see the Note in the Guide in documentation: https://registry.terraform.io/providers/jfrog/artifactory/latest/docs/guides/repositories_in_project#workaround

**Comment by railoni** [comment]
<img width="1728" alt="Screenshot 2024-01-31 at 11 03 59 AM" src="https://github.com/jfrog/terraform-provider-artifactory/assets/121967829/f5705b0c-02a0-44e7-bb7d-5de401aaa5d7">
<img width="1595" alt="Screenshot 2024-01-31 at 11 05 33 AM" src="https://github.com/jfrog/terraform-provider-artifactory/assets/121967829/4f99db8d-7435-4fde-9eff-686f0318eec8">
Here are code for our project creation and reference project in repos.
For federated our ha2 node project is not setting up. so we used local-exec to set project for it.
 using below code.
 provisioner "local-exec" {
    command = <<-EOT
      exec curl --location --request PUT '${var.artifactory-ha2_url}access/api/v1/projects/_/attach/repositories/${project.BU_project_HA2.key}-maven-shared-dev/${project.BU_project_HA2.key}?force=true/false' --header 'Authorization: Bearer ${var.artifactory-ha2_token}'
    EOT
  }

**Reply by alexhung** [reply]
There is no `repos` attribute for your `project` resource?

You have 2 options to manage the project <-> repos relationships:
1. Use `project_key` attribute in the repo resource
2. Use `repos` attribute in the project resource

Currently option 2 is the preferred/recommended way.

**Comment by railoni** [comment]
Do we need to Use repos attribute in the project resource.

specify all repos under project resource.

```terraform
resource "project" "myproject" {
  key          = "myproj"
  display_name = "My Project"
  description  = "My Project"
  admin_privileges {
    manage_members   = true
    manage_resources = true
    index_resources  = true
  }
  max_storage_in_gibibytes   = 10
  block_deployments_on_limit = false
  email_notification         = true

  repos = ["myproj-docker-v2-local"]

  depends_on = [ artifactory_local_docker_v2_repository.docker-v2-local ]
}
```

Could you please share some example for artifactory federated repository.

**Reply by alexhung** [reply]
As far as project is concerned, there's no difference between local, remote, virtual, or federated repos in the `repos` attribute. Include the federated repo's key in the list and it will be added to the project.
